File: apps/web_companion/src/api/social_api.ts
import { PresenceStatus } from '@chirp/proto/social';
import { MsgID } from '@chirp/proto/gateway';
import { ADD_FRIEND, FRIEND_REQUEST_ACTION, GET_PRESENCE, LOGIN, SET_PRESENCE } from '../protocol/msg_map';
import { setPresence } from '../state/presence_store';
import {
  addFriend,
  addPendingIn,
  addPendingOut,
  fromUserIdOf,
  loadFriendState,
  persistFriendStore,
  removeFriend,
  resolvePending,
  type FriendState,
} from '../state/friend_store';
import type { Store } from '../state/store';
import type { AuthState } from '../state/auth_store';
import type { PresenceState } from '../state/presence_store';
import type { ChatConnection } from './chat_api';
import { FriendAcceptedNotify, FriendRequestNotify, FriendRemovedNotify, PresenceNotify } from '@chirp/proto/social';

export interface SocialApiDeps {
  conn: ChatConnection;
  auth: Store<AuthState>;
  presence: Store<PresenceState>;
  friends: Store<FriendState>;
}

/**
 * Social plane api (second websocket, port 8001): friends and presence.
 * Degradable by design — when the social socket is unavailable the chat
 * keeps working and the UI simply hides friend features.
 */
export class SocialApi {
  private readonly conn: ChatConnection;
  private readonly auth: Store<AuthState>;
  private readonly presence: Store<PresenceState>;
  private readonly friends: Store<FriendState>;
  private unsubs: Array<() => void> = [];
  private persistOff: (() => void) | null = null;

  constructor(deps: SocialApiDeps) {
    this.conn = deps.conn;
    this.auth = deps.auth;
    this.presence = deps.presence;
    this.friends = deps.friends;
  }

  async login(userId: string): Promise<boolean> {
    if (this.conn.status !== 'connected') {
      await this.conn.connect();
    }
    // Same LOGIN pair as chat (scaffold token = user id); the social
    // service keeps no device registry, so no device_id is needed.
    const resp = await this.conn.request(LOGIN, { token: userId, platform: 'web' });
    if (resp.code !== 0) {
      // Degrade quietly: no friend features, chat unaffected.
      this.conn.disconnect();
      return false;
    }
    this.conn.resetBackoff();
    this.friends.set(() => loadFriendState(userId));
    this.persistOff = persistFriendStore(this.friends, userId);
    this.start();
    return true;
  }

  logout(): void {
    this.stop();
    this.conn.disconnect();
  }

  stop(): void {
    for (const off of this.unsubs) off();
    this.unsubs = [];
    this.persistOff?.();
    this.persistOff = null;
  }

  /** Advertise our presence; the service broadcasts it to our friends. */
  async setPresence(status: PresenceStatus, statusMessage = ''): Promise<void> {
    const userId = this.auth.get().userId;
    if (!userId) return;
    await this.conn.request(SET_PRESENCE, { userId, status, statusMessage });
  }

  /** Bulk pull for the visible roster (conversation peers, friends). */
  async pullPresence(userIds: string[]): Promise<void> {
    if (userIds.length === 0) return;
    const resp = await this.conn.request(GET_PRESENCE, { userIds });
    if (resp.code !== 0) return;
    const at = Date.now();
    for (const presence of resp.presences ?? []) {
      setPresence(this.presence, presence.userId, presence.status, presence.statusMessage, at);
    }
  }

  /** Pending-request model: the other side must accept before we are friends. */
  async addFriend(targetUserId: string): Promise<number> {
    const userId = this.auth.get().userId;
    if (!userId) return -1;
    const resp = await this.conn.request(ADD_FRIEND, {
      userId,
      targetUserId,
      message: '',
    });
    if (resp.code === 0) addPendingOut(this.friends, targetUserId);
    return resp.code;
  }

  async respondRequest(requestId: string, accept: boolean): Promise<number> {
    const userId = this.auth.get().userId;
    if (!userId) return -1;
    const from = fromUserIdOf(this.friends.get(), requestId);
    const resp = await this.conn.request(FRIEND_REQUEST_ACTION, { userId, requestId, accept });
    if (resp.code !== 0) return resp.code;
    resolvePending(this.friends, requestId);
    // The ACCEPTED notify excludes the actor, so accept-side bookkeeping is local.
    if (accept && from) addFriend(this.friends, from);
    return resp.code;
  }

  private start(): void {
    if (this.unsubs.length > 0) return;
    this.unsubs = [
      this.conn.onNotify(MsgID.PRESENCE_NOTIFY, (body) => this.onPresenceNotify(body)),
      this.conn.onNotify(MsgID.FRIEND_REQUEST_NOTIFY, (body) => this.onFriendRequestNotify(body)),
      this.conn.onNotify(MsgID.FRIEND_ACCEPTED_NOTIFY, (body) => this.onFriendAcceptedNotify(body)),
      this.conn.onNotify(MsgID.FRIEND_REMOVED_NOTIFY, (body) => this.onFriendRemovedNotify(body)),
    ];
  }

  private onPresenceNotify(body: Uint8Array): void {
    let notify: PresenceNotify;
    try {
      notify = PresenceNotify.decode(body);
    } catch {
      return;
    }
    setPresence(this.presence, notify.userId, notify.status, notify.statusMessage, Date.now());
  }

  private onFriendRequestNotify(body: Uint8Array): void {
    let notify: FriendRequestNotify;
    try {
      notify = FriendRequestNotify.decode(body);
    } catch {
      return;
    }
    addPendingIn(this.friends, notify.requestId, notify.fromUserId);
  }

  private onFriendAcceptedNotify(body: Uint8Array): void {
    let notify: FriendAcceptedNotify;
    try {
      notify = FriendAcceptedNotify.decode(body);
    } catch {
      return;
    }
    // Server quirk: the ACCEPTED notify goes to the requester with user_id
    // set to the requester themselves, so the accepting peer's id is not
    // carried. Ignore self echoes (never friend ourselves); the requester's
    // roster gap is closed by the social backend work (P1 backlog).
    if (notify.userId === this.auth.get().userId) return;
    addFriend(this.friends, notify.userId);
  }

  private onFriendRemovedNotify(body: Uint8Array): void {
    let notify: FriendRemovedNotify;
    try {
      notify = FriendRemovedNotify.decode(body);
    } catch {
      return;
    }
    removeFriend(this.friends, notify.userId);
  }
}

File: apps/mobile_companion/lib/api/social_api.dart
import 'dart:typed_data';

import 'package:chirp_proto/chirp_proto.dart';

import '../protocol/chat_connection.dart';
import '../protocol/chirp_client.dart';
import '../protocol/msg_map.dart' as specs;
import '../state/auth_store.dart';
import '../state/friend_store.dart';
import '../state/store.dart';
import '../state/typing_presence.dart';

/// Social plane api (second websocket, port 8001): friends and presence.
/// Degradable by design — when the social socket is unavailable the chat
/// keeps working and the UI simply hides friend features. The backend owns
/// the roster: login pulls friends and pending requests, notifications keep
/// this mirror in sync.
class SocialApi {
  SocialApi({
    required this.conn,
    required this.auth,
    required this.presence,
    required this.friends,
  });

  final ChatConnection conn;
  final AuthStore auth;
  final Store<PresenceState> presence;
  final Store<FriendState> friends;
  final List<void Function()> _unsubs = [];

  Future<bool> login(String userId) async {
    if (conn.status != ConnStatus.connected) {
      await conn.connect();
    }
    // Same LOGIN pair as chat (scaffold token = user id); device_id stays
    // unset and lands on the service's "default" device.
    final resp = await conn.request(
        specs.login, LoginRequest(token: userId, platform: 'android'));
    if (resp.code != ErrorCode.OK) {
      // Degrade quietly: no friend features, chat unaffected.
      conn.disconnect();
      return false;
    }
    conn.resetBackoff();
    // Reset the local mirror (a previous account's roster must not leak),
    // bring notify handlers up first so nothing racing the pull is lost,
    // then load the authoritative roster.
    friends.value =
        const FriendState(friends: [], pendingIn: [], pendingOut: []);
    start();
    await fetchFriends();
    await fetchPending();
    return true;
  }

  void logout() {
    stop();
    conn.disconnect();
  }

  void stop() {
    for (final off in _unsubs) {
      off();
    }
    _unsubs.clear();
  }

  /// Re-pull the authoritative friend list (e.g. after a reconnect).
  Future<int> fetchFriends() async {
    final userId = auth.value.userId;
    if (userId == null) return -1;
    final resp = await conn.request(
        specs.getFriendList,
        GetFriendListRequest(
          userId: userId,
          limit: 0,
          offset: 0,
        ));
    if (resp.code != ErrorCode.OK) return resp.code.value;
    replaceFriends(friends, resp.friends.map((f) => f.userId).toList());
    return resp.code.value;
  }

  /// Re-pull the incoming request queue.
  Future<int> fetchPending() async {
    final userId = auth.value.userId;
    if (userId == null) return -1;
    final resp = await conn.request(
        specs.getPendingRequests, GetPendingRequestsRequest(userId: userId));
    if (resp.code != ErrorCode.OK) return resp.code.value;
    replacePendingIn(
        friends,
        resp.requests
            .map((r) => PendingRequest(
                requestId: r.requestId, fromUserId: r.fromUserId))
            .toList());
    return resp.code.value;
  }

  /// Advertise our presence; the service broadcasts it to our friends.
  Future<void> publishPresence(PresenceStatus status,
      {String statusMessage = ''}) async {
    final userId = auth.value.userId;
    if (userId == null) return;
    await conn.request(
        specs.setPresence,
        SetPresenceRequest(
          userId: userId,
          status: status,
          statusMessage: statusMessage,
        ));
  }

  /// Bulk pull for the visible roster (conversation peers, friends).
  Future<void> pullPresence(List<String> userIds) async {
    if (userIds.isEmpty) return;
    final resp = await conn.request(
        specs.getPresence,
        GetPresenceRequest(
          userIds: userIds,
        ));
    if (resp.code != ErrorCode.OK) return;
    final at = DateTime.now().millisecondsSinceEpoch;
    for (final entry in resp.presences) {
      setPresence(
          presence, entry.userId, entry.status, entry.statusMessage, at);
    }
  }

  /// Pending-request model: the other side must accept before we are friends.
  Future<int> sendFriendRequest(String targetUserId) async {
    final userId = auth.value.userId;
    if (userId == null) return -1;
    final resp = await conn.request(
        specs.addFriend,
        AddFriendRequest(
          userId: userId,
          targetUserId: targetUserId,
          message: '',
        ));
    if (resp.code == ErrorCode.OK) addPendingOut(friends, targetUserId);
    return resp.code.value;
  }

  Future<int> respondRequest(String requestId, bool accept) async {
    final userId = auth.value.userId;
    if (userId == null) return -1;
    final resp = await conn.request(
        specs.friendRequestAction,
        FriendRequestAction(
          userId: userId,
          requestId: requestId,
          accept: accept,
        ));
    if (resp.code != ErrorCode.OK) return resp.code.value;
    resolvePending(friends, requestId);
    // Roster sync arrives via the ACCEPTED notify, which the server sends to
    // both sides carrying the OTHER party's id.
    return resp.code.value;
  }

  /// Symmetric, idempotent delete on the server; the peer learns via notify.
  Future<int> removeFriend(String targetUserId) async {
    final userId = auth.value.userId;
    if (userId == null) return -1;
    final resp = await conn.request(
        specs.removeFriend,
        RemoveFriendRequest(
          userId: userId,
          friendUserId: targetUserId,
        ));
    if (resp.code == ErrorCode.OK) removeFriendStore(friends, targetUserId);
    return resp.code.value;
  }

  /// Blocking severs the friendship on the server (both ways).
  Future<int> blockUser(String targetUserId) async {
    final userId = auth.value.userId;
    if (userId == null) return -1;
    final resp = await conn.request(
        specs.blockUser,
        BlockUserRequest(
          userId: userId,
          targetUserId: targetUserId,
        ));
    if (resp.code == ErrorCode.OK) removeFriendStore(friends, targetUserId);
    return resp.code.value;
  }

  Future<int> unblockUser(String targetUserId) async {
    final userId = auth.value.userId;
    if (userId == null) return -1;
    final resp = await conn.request(
        specs.unblockUser,
        UnblockUserRequest(
          userId: userId,
          targetUserId: targetUserId,
        ));
    return resp.code.value;
  }

  void start() {
    if (_unsubs.isNotEmpty) return;
    _unsubs.addAll([
      conn.onNotify(MsgID.PRESENCE_NOTIFY, onPresenceNotify),
      conn.onNotify(MsgID.FRIEND_REQUEST_NOTIFY, onFriendRequestNotify),
      conn.onNotify(MsgID.FRIEND_ACCEPTED_NOTIFY, onFriendAcceptedNotify),
      conn.onNotify(MsgID.FRIEND_REMOVED_NOTIFY, onFriendRemovedNotify),
    ]);
  }

  void onPresenceNotify(Uint8List body) {
    PresenceNotify notify;
    try {
      notify = PresenceNotify.fromBuffer(body);
    } catch (_) {
      return;
    }
    setPresence(presence, notify.userId, notify.status, notify.statusMessage,
        DateTime.now().millisecondsSinceEpoch);
  }

  void onFriendRequestNotify(Uint8List body) {
    FriendRequestNotify notify;
    try {
      notify = FriendRequestNotify.fromBuffer(body);
    } catch (_) {
      return;
    }
    addPendingIn(friends, notify.requestId, notify.fromUserId);
  }

  void onFriendAcceptedNotify(Uint8List body) {
    FriendAcceptedNotify notify;
    try {
      notify = FriendAcceptedNotify.fromBuffer(body);
    } catch (_) {
      return;
    }
    // user_id is always the OTHER party of the new friendship. A self echo
    // would mean "friend ourselves" — impossible from the current server,
    // but the guard is cheap and keeps a broken peer from poisoning us.
    if (notify.userId == auth.value.userId) return;
    addFriend(friends, notify.userId);
  }

  void onFriendRemovedNotify(Uint8List body) {
    FriendRemovedNotify notify;
    try {
      notify = FriendRemovedNotify.fromBuffer(body);
    } catch (_) {
      return;
    }
    removeFriendStore(friends, notify.userId);
  }
}
